SPI API cleanup and update

tbursztyka opened the pull request:
Please review

The patch-set disables the legacy api by default, applies the feedback and then move most of the code using legacy API to the new API.

Some other patch might come still, (esp ethernet stuff for instance) but it's worth reviewing the patches already.

Commit:
drivers/spi: Add support for TX or RX only modes on DW driver

As for RX only, computing the NDF will be used for EEPROM mode.
Only a way to determine EEPROM mode is missing.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Cleanup the Kconfig files

Split QMSI relevant part into its own file. Some config where using
prompt, some not: normalizing it by removing the prompt keyword where
relevant.

Reducing the file by using if/endif when relevant.

However, it still not fully clean default: cfg and default baudrate
should disappear. There is no default configuration to apply as long
as the controller is not configured to run from any part using spi API.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Adapt Kconfig and generic context to enable slave support

Adding Kconfig options to set supported modes by the controller
(master, slave or both)

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Add slave mode support to the DesignWare driver

It involves a minor change on which register is configured.
Most of the change is with threshold handling.
Handling the Kconfig based supported mode per-port.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

locomuco committed:
drivers/ethernet: Switch enc28j60 to new SPI API

Let's use the new SPI API and ditch the old one.

Signed-off-by: Matthias Boesl <matthias.boesl@gmail.com>
Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/ieee802154: Switch CC2520 to new SPI API

Finally switching to new SPI API. It makes code simpler and performs
better.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/quark_se: Switch to native SPI DW driver

QMSI is not updated to latest SPI API. Switching to native DW driver,
until we find a way either to update the shim drivers in a
non-performant way, or updating QMSI itself enabling support for
scatter-gather type of buffers.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
boards/x86: Pinmux SPI port 2 relevantly on quark_se_c1000_devboard

This port is a slave SPI port only.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
api/spi: Make spi_config parameter constant

As the content of this struct will not be modified by drivers, it's
better to pass it as constant. Also, if someday struct device can be
made contant too, this change will make spi ready for registering the
spi_config into ROM directly.

Fixes #5839

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
api/spi: Add a dedicated Kconfig option for asynchronous mode enablement

Instead of using CONFIG_POLL, which is not directly related to SPI and
is a kernel option, let's have SPI_ASYNC instead. When enabled, it will
select POLL automatically.

Fixes #5839

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
api/spi: Removing eeprom mode in configuration

EEPROM mode can be, in fact, asserted from tx_bufs/rx_bufs thus removing
it.

Fixes #5839

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
api/spi: Reduce parameter number on transceive function

tx_bufs/tx_count and rx_bufs/rx_count can be hold in another dedicated
structure, thus reducing the number of parameters to transceive. This
permits to avoid using the stack when calling transceive.

Since we saved parameters, we can expose back the struct device pointer,
to stay consistent with other device driver APIs.

Fixes #5839

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/flash: Switch W25QXXDV driver to new SPI API

This saves quite some memory.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/sensors: Switch bme280 driver to new SPI API

Replacing legacy API calls by news ones.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/sensors: Switch bmi160 driver to new SPI API

Replacing legacy API calls by news ones.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/sensors: Switch adxl362 driver to new SPI API

Replacing legacy API calls by news ones.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/sensors: Switch lis2dh driver to new SPI API

Replacing legacy API calls by news ones.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/adc: Switch ti_adc108s102 driver to new SPI API

Replacing legacy API calls by news ones.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/bluetooth: Switch SPI based HCI driver to new SPI API

Replacing legacy API calls by news ones.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
api/spi: Disable legacy API by default

Let's start deprecation work of the SPI legacy API.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

carlescufi commented:
Sorry @tbursztyka I assigned before I read :) @anangl please don't review yet

tbursztyka commented:
@carlescufi no problem, a lot more commits are coming, so reviewer list will be long then.

locomuco commented:
@tbursztyka this one might be related, just in case... 
https://github.com/zephyrproject-rtos/zephyr/pull/662

tbursztyka commented:
@locomuco yes, good point, I'll backport it

vanwinkeljan commented:
@tbursztyka fyi (rfc) pull request #5670 takes care of moving nrf52 master spi to new API

Commit:
samples/drivers: Switch Fujistu FRAM sample to new SPI API

Replacing legacy API calls by news ones.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
Finally check passed...  However, as it's only about compilation (I could not test most of the hw drivers updated), it will require thorough review.

Commit:
drivers/bluetooth: Get rid completely of legacy SPI API in SPI HCI

No need to support legacy API there.
The special usage of CS in BlueNRG is centralized it dedicated functions
to avoid all the #ifdefs everywher.
Also managing transaction error.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

ydamigos commented on drivers/spi/Kconfig.stm32:
```diff
@@ -0,0 +1,30 @@
+menuconfig SPI_STM32
+	bool "STM32 MCU SPI controller driver"
+	depends on SPI && SOC_FAMILY_STM32
+	depends on SOC_SERIES_STM32L4X || SOC_SERIES_STM32F4X || SOC_SERIES_STM32F3X
```
Why did you leave out the previous dependencies and selections?
```
depends on SPI && SOC_FAMILY_STM32
depends on SOC_SERIES_STM32L4X || SOC_SERIES_STM32F0X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32F4X
select HAS_DTS_SPI
select USE_STM32_LL_SPI
```

tbursztyka replied:
looks like one of the many rebase I did that missed this.

tbursztyka replied:
Actuall, that explains the DTS issue as well, since select HAS_DTS_SPI was missing ...

ydamigos commented on boards/arm/disco_l475_iot1/Kconfig.board:
```diff
@@ -8,5 +8,6 @@
 config BOARD_DISCO_L475_IOT1
 	bool "Discovery IoT L475 Development Board"
 	select HAS_DTS_I2C_DEVICE
+	select HAS_DTS_SPI
```
Why should we enable HAS_DTS_SPI per board? When SPI_STM32 is enabled, it could select HAS_DTS_SPI in Kconfig.

ydamigos commented on boards/arm/96b_carbon/Kconfig.board:
```diff
@@ -7,5 +7,6 @@
 
 config BOARD_96B_CARBON
 	bool "96Boards Carbon (STM32F401)"
+	select HAS_DTS_SPI
```
Why should we enable HAS_DTS_SPI per board? When SPI_STM32 is enabled, it could select HAS_DTS_SPI in Kconfig.

tbursztyka replied:
So it's fine to select HAS_DTS_SPI from the config HAS_DTS_SPI_DEVICE? First time I mess with dts.

ydamigos replied:
I was not clear. I mean to select HAS_DTS_SPI in `menuconfig SPI_STM32`. We use HAS_DTS_SPI_DEVICE for the devices connected on the SPI bus (e.g. sensors, bluetooth modules).

tbursztyka replied:
yes, realized that while fixing the rebase issue

ydamigos reviewed (commented):
Some comments regarding the new Kconfig.stm32.

anangl commented on include/spi.h:
```diff
@@ -102,6 +102,7 @@ extern "C" {
 #define SPI_LINES_SINGLE	(0)
 #define SPI_LINES_DUAL		BIT(11)
 #define SPI_LINES_QUAD		BIT(12)
+#define SPI_LINES_OCTAL		BIT(13)
```
Bit 13 is still used by `SPI_HOLD_ON_CS`, only the documentation of `spi_config` has been changed, not the bit definitions themselves.
But wouldn't it be better to stay with bits [11 : 12] assigned to SPI_LINES configuration, and define the options as below?
```
#define SPI_LINES_SINGLE	(0 << 11)
#define SPI_LINES_DUAL		(1 << 11)
#define SPI_LINES_QUAD		(2 << 11)
#define SPI_LINES_OCTAL		(3 << 11)
```
They cannot be selected together.

tbursztyka replied:
True, it's like the word size. Missed that

anangl commented on drivers/spi/spi_context.h:
```diff
@@ -20,19 +20,19 @@ extern "C" {
+	const struct spi_buf *current_rx;
 	size_t rx_count;
 
 	u8_t *tx_buf;
```
This can be `const u8_t *`.

anangl commented:
One more improvement suggestion.
In the following part of `spi_context_update_tx` (in `spi_context_update_rx` it is similar):
```
 	if (!ctx->tx_len) {
 		ctx->current_tx++;
 		ctx->tx_count--;
 
 		if (ctx->tx_count) {
```
the incrementation of `ctx->current_tx` can be moved inside `if (ctx->tx_count) {`. There is no need to increase the pointer after last buffer is sent.

tbursztyka commented on drivers/spi/spi_handlers.c:
```diff
@@ -27,19 +27,33 @@ static void verify_spi_buf_array(struct spi_buf *bufs, size_t len,
-	verify_spi_buf_array((struct spi_buf *)rx_bufs, rx_count, 1, ssf);
+	if (tx_bufs) {
+		const struct spi_buf_set *tx =
+			(const struct spi_buf_set *)tx_bufs;
```
because tx_bufs is a u32_t if I remember well.

anangl replied:
Right. It's `_SYSCALL_HANDLER`. I somehow overlooked this.

tbursztyka commented on drivers/spi/spi_handlers.c:
```diff
@@ -27,19 +27,33 @@ static void verify_spi_buf_array(struct spi_buf *bufs, size_t len,
+			(const struct spi_buf_set *)tx_bufs;
+
+		_SYSCALL_MEMORY_READ(tx_bufs, sizeof(struct spi_buf_set));
+		verify_spi_buf_array((struct spi_buf *)tx->buffers,
```
true

tbursztyka commented on drivers/spi/spi_sam0.c:
```diff
@@ -312,27 +328,40 @@ static void spi_sam0_fast_transceive(struct spi_config *config,
+	size_t rx_count;
+
+	if (!tx_bufs || !rx_bufs) {
+		return false;
```
wrong bitwise test indeed.

anangl commented:
The `cs` member of the `spi_config` structure can be also defined as pointing to a constant structure, i.e. `const struct spi_cs_control *cs`.

Commit:
api/spi: Add octal MISO lines mode

Some controllers can provide up to 8 MISO lines.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
api/spi: Make cs attribute in struct spi_config constant

Content is never modified by any driver.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

ydamigos commented:
BlueNRG doesn't work correctly with the recent changes to bluetooth SPI HCI due to CS. It needs to hold CS low between some SPI transactions.

Does `struct spi_cs_control` support devices with active high CS?

tbursztyka commented:
@ydamigos CS needs to hold on between transactions? If so, the bit cs_hold in spi_config should help. You can switch it on/off between transactions. Let me fix the patch, you'll tell me if that works.

ydamigos commented:
@tbursztyka I already send you a patch directly to your branch. I tried to fix it using cs_hold bit but it is not straight forward. The code checks if the BlueNRG module is awake by checking the SPI response. If it is awake it should hold the CS. If it is not it should release it and retry. If we set the cs_hold bit for the operation, could we override it somehow if the module is not ready?

tbursztyka commented:
@ydamigos  hum did not see the patch? well I updated the patch, tell me if that's ok.
Does bluerng has a delay for starting? you could play on spi_cs_control's delay attribute. It will wail the delay after setting up the CS low, and then wait again before setting it high. Would that help?
If find it dangerous to have these do {} while, it could loop forever.

tbursztyka commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -318,13 +285,10 @@ static int bt_spi_send(struct net_buf *buf)
 
 	/* Transmit the message */
 	do {
+		spi_conf.operation &= &SPI_HOLD_ON_CS;
```
crap...  extra &

ydamigos commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -203,24 +175,21 @@ static void bt_spi_rx_thread(void)
 
 		size = header_slave[STATUS_HEADER_TOREAD];
 		do {
+			spi_conf.operation &= SPI_HOLD_ON_CS;
```
`~SPI_HOLD_ON_CS`?

tbursztyka replied:
oh yes...

ydamigos commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -300,12 +269,10 @@ static int bt_spi_send(struct net_buf *buf)
 		return -EINVAL;
 	}
 
+	spi_conf.operation |= SPI_HOLD_ON_CS;
```
`~SPI_HOLD_ON_CS`?

ydamigos replied:
Not here. Below to the `&=`

tbursztyka replied:
I am bit lost: before it was setting CS low, then trying to enable the chip, if so it kept it low, transceive, then putting it high. 

That's what I tried to do with cs_hold bit: on enablement it does not set CS high again, it's kept low. On next transceive it will put it high again.

ydamigos replied:
I was not clear. My second comment refers to my first comment. I meant to add the first comment on line 288 and not on line 272.

ydamigos reviewed (commented):
@tbursztyka The patch is here: https://github.com/tbursztyka/zephyr/pull/1. I will check it if it could work with spi_cs_control's delay.

tbursztyka commented:
@ydamigos  github did not send me any notification on your patch. Anyway, if playing with cs_hold and delay is not solving anythin, I'll revert to direct gpio manipulation as you did.

anangl commented on include/spi.h:
```diff
@@ -141,6 +137,32 @@ struct spi_cs_control {
+	/** When slave chip is deselected */
+	SPI_SLAVE_EVENT_DESELECTED	= 0,
+	/** When slave chip is selected */
+	SPI_SLAVE_EVENT_SELECTED,
```
It may be difficult, or nearly impossible, to provide these two events on some controllers. nRF5 family will be such case.
Do we really need them?

tbursztyka replied:
If slave CS is deferred to gpio, it will be useful. Not all slave will need these but some might.

anangl replied:
If these two events are optional, it should be documented. So that no one expects an event that will never appear.

superna9999 replied:
I don't really see how the controller can send the SPI_SLAVE_EVENT_INPUT event ? how do the controller knows that ?
For the SPI_SLAVE_EVENT_ERROR, this should be returned by the transceive call, no ?

tbursztyka replied:
Documentation is too thin indeed.

About the error event I am not entirely decided either (as for the callback returning an int, see below): let's say there is a driver issue, we could of course return an error code on the spi_transceive() (the very first one which is waiting for sync), and it will actually do that. But what about the slave callback and the slave's state machine logic (if any, some slave will be super dumb, some will be quite complicated): maybe it's worth notifying the error up-front via the callback so slave code could cleanup things etc... 
Let's see, maybe I am over-complexifying the slave user part, and up to the dev to handle the error code only for spi_transceive and cleanup/reset what's necessary.

Commit:
api/spi: Add the possibility to request CS active high logic

Some device may need to be put up on CS high logic. The active low logic
is the default as usual, but it is now possible to request the active
high logic.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

ydamigos commented:
> Anyway, if playing with cs_hold and delay is not solving anything, I'll revert to direct gpio manipulation as you did.

@tbursztyka I tried with cs_hold and delay without success.

tbursztyka commented:
@ydamigos even with SPI_CS_ACTIVE_HIGH config bit?  I think I have a bluenrg somewhere. I might try it, so I could try to verify this use case. It's the only one so far that could make use of delay, cs_hold and this cs active high bit.

ydamigos commented:
@tbursztyka SPI_CS_ACTIVE_HIGH was for another IC, which, I found out, is broken so I cannot test it. The bluenrg works with active low CS.

tbursztyka commented:
@ydamigos ok. So I do have 2 bluenrg (nucleo idb05a1), I'll test it. Actually, I missed the hci_spi samples: it still has legacy stuff there.

tbursztyka commented:
@ydamigos  ok it's corrected. @erwango provided me the right datasheet for bluenrg, and indeed it's behavior is particular with CS. I made the code a tiny bit cleaner, and managed transaction error (if any). It seems to work properly. Try it, and let me know if I missed anything.

ydamigos commented:
@tbursztyka Thanks for the changes it works great now.
I see you also removed the SPI HCI sample. As I understand it, this sample is used to make a SoC (e.g. nRF51 on 96b_carbon) to act as a Bluetooth controller, like BlueNRG, and the HCI SPI driver acts as the Host part of the Bluetooth HCI interface. Is it ok to remove it?

erwango commented:
> Is it ok to remove it?

Agree with @ydamigos , this sample is required to build BLE controller.
Though, maybe it could be renamed to ble_controller, since obviously current name is confusing

tbursztyka commented:
@locomuco I backported your patch into this PR. You can close your PR #662 then. I merged the 2 patches at once since there is no need to support both API now, and I of course applied the changes brought in that PR (and fixed gpio cs handling: a dedicated boolean Kconfig was necessary). But the driver is far from using the possibility of the new API: it's still doing memory copies everywhere, with statically allocated buffer, not centralizing the spi access also (which I think in this driver would be easy to do). We'll need to get back to it at some point.

@anangl I pushed your nrf shim into this PR as well.

tbursztyka commented:
@erwango I put it back yes. It seems to be working only on 92b_carbon board, and with the legacy API. My take is that it will need to be updated once the spi slave API is proven to work and drivers updated as well.

anangl commented:
@tbursztyka But the commit was not well prepared for merging. I think that at least the changes done to the "spi_loopback" test should be extracted to a separate one.

tbursztyka commented:
@anangl indeed, missed that part (I just stared at the driver code which looked fine to me).

vanwinkeljan commented on drivers/spi/spi_nrf5.c:
```diff
@@ -0,0 +1,481 @@
+	const struct spi_nrf5_cfg *dev_cfg = DEV_CFG(spi_cfg);
+	u8_t slave = spi_cfg->slave;
+
+	nrfx_spi_cfg->sck_pin  = dev_cfg->psel.sck;
```
The GPIO pins for the SPI should only be configured once during init on not per configuration change

vanwinkeljan commented on drivers/spi/spi_nrf5.c:
```diff
@@ -0,0 +1,481 @@
+
+static inline nrfx_spi_frequency_t spi_freq_to_nrfx_freq(u32_t frequency)
+{
+	switch (frequency) {
```
Would it not be better to use closed matching frequency instead of expecting the caller to give the exact frequency.

vanwinkeljan commented on drivers/spi/Kconfig.nrf5:
```diff
@@ -0,0 +1,293 @@
+	int
+	default 255
+
+endif # SPI_NRF5
```
Missing SPI 2 & 3

vanwinkeljan commented on drivers/spi/spi_nrf5.c:
```diff
@@ -0,0 +1,481 @@
+		return NRFX_SPI_FREQ_4M;
+	case 8000000:
+		return NRFX_SPI_FREQ_8M;
+	default:
```
Missing frequencies for nrf52840 ab (16M & 32M)

vanwinkeljan commented on drivers/spi/spi_nrf5.c:
```diff
@@ -0,0 +1,481 @@
+static inline void set_buffers_len_for_nrfx(nrfx_spi_xfer_desc_t *xfer_desc,
+					    u32_t len)
+{
+	if (xfer_desc->p_tx_buffer) {
```
Is it somewhere taken in account that the length can not be larger then 255 for nrf58240 aa?

Commit:
tests/drivers: Removing old SPI test

It was much more limited than the spi_loopback test, and of course using
the legacy API. Removing it then.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

dbkinder reviewed (commented):
no doc changes so removing myself from reviewers

galak commented on include/spi.h:
```diff
@@ -152,25 +183,32 @@ struct spi_cs_control {
+#else
 	u32_t		frequency;
+#endif /* CONFIG_SPI_SLAVE */
 	u16_t		operation;
```
It seems like we need to be able to support SPI slave & master at the same time.  I could see a device w/multiple SPI controllers with one connected to sensors(master) and another to a host processor (slave).

tbursztyka replied:
Some controllers do support both at same time, some only one of the 2 modes. 
That's a driver issue. It's already taken care of in spi_context (in a patch which is not here actually, I'll update that). It really does not need to be exposed in the user API itself.

tbursztyka replied:
Actually, the patch was missing 4 important lines, that got lost in one rebase: the enum that dictates runtime mode support (master/slave). Up to the device drivers then to use it (dw driver will be updated along with slave support). As it is not part of the public API, it could be changed/improved.

galak replied:
Sounds good, just don't want us to preclude the use model I suggested.

galak reviewed (changes requested):
We should be able to support slave & master mode in the same build, maybe having 2 different controller instances with one as master and one as slave.

MaureenHelm committed:
spi: Implement new spi api in the mcux dspi driver

Adds support for the new spi api to the mcux dspi shim driver. Does not
remove support for the legacy spi api since there are still consumers of
that api, particularly the mcr20a 802.15.4 driver.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
tests: spi_loopback: Add frdm_k64f configuration

Adds a board-specific configuration for the frdm_k64f to the
spi_loopback test.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

superna9999 commented on include/spi.h:
```diff
@@ -386,6 +419,78 @@ static inline int _impl_spi_release(const struct spi_config *config)
+ *
+ * @retval 0 If successful, negative errno code otherwise.
+ */
+static inline int spi_slave_update_buffers(const struct spi_config *config,
```
I don't understand the point to have the same spi_slave_transceive & spi_slave_update_buffers which are the exact same as spi_transceive ? adding the events should be enough I think.

tbursztyka replied:
It's just aliases, for clearer user code. But indeed, nothing prevents you to just use spi_transceive directly.

superna9999 replied:
In this case, why don't you use the spi_transceive() function to be able to drive it from an application ?

tbursztyka replied:
I don't follow you. You can use spi_transceive as well on a spi slave application, I just made these aliases for clarity. I still have to finish testing slave support on dw, but on the sample side, with the slave cb it's obvious when you read those aliases what is being done.

superna9999 replied:
Inline aliases won't hurt anyway.

tbursztyka replied:
Yes, actually these are like spi_write/spi_read. Better wording would be "helpers", instead of aliases. Maybe it's worth documenting it.

superna9999 commented:
Apart the useless spi_slave_ functions, Ok for me.

tbursztyka commented:
@anangl could you re-review?

anangl commented on drivers/spi/spi_context.h:
```diff
@@ -113,78 +128,140 @@ static inline void spi_context_complete(struct spi_context *ctx, int status)
+			gpio_remove_callback(ctx->config->cs->gpio_dev,
+					     &ctx->cs_cb);
+		}
+	}
```
Missing `return` at the end of this block, I guess.

tbursztyka replied:
Indeed

anangl commented on include/spi.h:
```diff
@@ -152,25 +183,32 @@ struct spi_cs_control {
  *
- * @note cs_hold, lock_on and eeprom_rx can be changed between consecutive
+ * @note cs_hold and lock_on can be changed between consecutive
  * transceive call.
```
This note is a bit confusing. Does it imply that other fields of `operation` cannot be changed in consecutive calls?

tbursztyka replied:
Yes, and it should be specifically said here for clarity.

anangl replied:
But why? Why couldn't a controller perform one transfer in e.g. mode 0 with word size 8, then another one (perhaps with a different slave but not necessarily) in mode 3 with word size 16?

tbursztyka replied:
It totally could. But then it means you are not accessing the same slave, is it's another spi_config you need for this.

I mean, I don't see any real use-case where you would be able to switch word-size when dealing with one specific slave. Does that exist?

anangl replied:
I saw such behavior when a CPU booted from an external flash - first it was trying to read an initial part of the boot image with low frequency and 8 bit word size. When this was successful, it switched to higher frequency and 16-bit words (and changed SPI mode from 0 to 3 as well).
And I can see this as a use case. If you don't know exactly what slave (with what capabilities) you can expect on the bus, first you do a transaction with some "safe" configuration just to collect required info, then you switch settings to the ones providing the best performance.

tbursztyka replied:
This is a really particular use case, which could be addressed by having 2 spi_config: safe one vs performance one. I am not sure we want to generalize this. Because going through the controller configuration at every transaction means quite a lot of cpu cycles: reading/writing registers etc... Of course we could make controller copying the config directly, and it would do a memcmp, but that's also a little bit under-performing and takes a bit of memory.

Imo, for thus case you describe, 2 spi_config would do.

anangl commented on include/spi.h:
```diff
@@ -141,6 +146,32 @@ struct spi_cs_control {
+ * @param config Pointer to the current slave spi_config structure instance.
+ * @param event
+ */
+typedef int (*spi_slave_cb)(const struct spi_config *config,
```
What is the purpose of the return value?

tbursztyka replied:
I am not entirely decided on that, but here is the idea: let's say you get an input event, and in your slave code you check what the master has sent you so far. And it's full garbage: shall we notify to the driver so it can stop the transaction by returning an error code?

anangl commented on include/spi.h:
```diff
@@ -116,11 +117,15 @@ extern "C" {
- * the mode and thus should never return -EINVAL on that configuration)
+
+/* Active high logic on CS - Usually, and by default, CS logic is active
+ * low. However, some devices may require the reverse logice: active high.
```
typo: logic

anangl commented on include/spi.h:
```diff
@@ -152,25 +183,32 @@ struct spi_cs_control {
  *     cs_hold             [ 13 ]      - Hold on the CS line if possible.
  *     lock_on             [ 14 ]      - Keep resource locked for the caller.
- *     eeprom              [ 15 ]      - EEPROM mode.
+ *     cs_active_high      [ 15 ]      - Active high CS logice.
```
typo: logic

anangl commented on include/spi.h:
```diff
@@ -395,14 +419,86 @@ static inline int spi_write_async(struct spi_config *config,
+ * Note: If this is the only mode supported by the controller, it will just
+ *       put the controller to sleep and it will not resond anymore to any
+ *       data sent to it by a master. If case both modes are supported at
+ *       runtime by the controller, it will be the same but additionnaly
```
additionally

anangl commented on include/spi.h:
```diff
@@ -395,14 +419,86 @@ static inline int spi_write_async(struct spi_config *config,
+ *
+ * Note: If this is the only mode supported by the controller, it will just
+ *       put the controller to sleep and it will not resond anymore to any
+ *       data sent to it by a master. If case both modes are supported at
```
In case

tbursztyka replied:
ouch ouch. and  resond/respond.

anangl commented on drivers/spi/spi_sam0.c:
```diff
@@ -318,27 +335,40 @@ static void spi_sam0_fast_transceive(struct spi_config *config,
+	size_t rx_count;
+
+	if (!tx_bufs && !rx_bufs) {
+		return false;
```
It seems that there is something wrong here. When both `tx_bufs` and `rx_bufs` are `NULL` the function returns immediately but when only one of them is `NULL`, this pointer is dereferenced below.

tbursztyka commented on drivers/spi/spi_context.h:
```diff
@@ -113,78 +128,140 @@ static inline void spi_context_complete(struct spi_context *ctx, int status)
+			   ctx->config->cs->gpio_pin,
+			   GPIO_DIR_OUT |
+			   (spi_context_cs_inactive_value(ctx) ?
+			    GPIO_PUD_PULL_UP : GPIO_PUD_PULL_DOWN));
```
No that's for the newly added SPI_CS_ACTIVE_HIGH bit in spi_config's operation attribute.

The input part for slave, when the slave's CS logic is deferred to a GPIO line, is just above that part.

anangl replied:
But you are enabling pull-up or pull-down on an output pin here. Was that really the intention?

tbursztyka replied:
nrf5 spi needs this, maybe others I don't know. Well, maybe I'll remove that, up to these specific driver to do what's necessary then.

(Actually, it's missing on slave part also).

anangl replied:
We've got some misunderstanding here. nrf5 for sure does not need this.

tbursztyka replied:
Ok, so drivers/spi/spim_nrf52_legacy.c was doing something wrong ?

Well, in any case good that you pointed out this: I removed it entirely from spi_context.h

Commit:
api/spi: Precise a bit the documentation

Notify about helpers

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/quark_se: Enable SPI port 2 as a slave only

Quark SE provides a slave SPI.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

mbolivar reviewed (approved):
LED strip changes seem fine.

One random question about the way this interacts with DTS, probably not relevant.

tbursztyka commented on drivers/spi/Kconfig:
```diff
@@ -98,250 +46,185 @@ config SYS_LOG_SPI_LEVEL
-	string
-	prompt "SPI port 0 device name"
-	depends on SPI_0
+	string "SPI port 0 device name"
```
sounds like a rebase collision to me, will fix it.

ydamigos replied:
>Also, my (limited) understanding of DTS is that the general practice is that the device will be compiled out completely unless the SPI DTS node has status = "ok";

@mbolivar Your understanding is correct but, right now, enabling SPIx port in dts does not enable CONFIG_SPI_x.

mbolivar replied:
> @mbolivar Your understanding is correct but, right now, enabling SPIx port in dts does not enable CONFIG_SPI_x.

@ydamigos, thanks for your response. So enabling CONFIG_SPI_x via DTS is more of a TODO item?

ydamigos replied:
>So enabling CONFIG_SPI_x via DTS is more of a TODO item?

Yes and other devices also (e.g. I2C).

Commit:
drivers/spi: Switch Intel driver to new SPI API

Ditch any legacy API support altogether.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Remove legacy DesignWare SPI driver

Nothing requires this driver anymore.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

anangl commented on include/spi.h:
```diff
@@ -152,25 +201,33 @@ struct spi_cs_control {
- * transceive call.
+ * @note Only cs_hold and lock_on can be changed between consecutive
+ * transceive call. Rest of the attributes are not meant to be tweaked.
+ * (I.e.: one struct spi_config is necessary per-device on the bus)
```
Once more regarding not changing the other fields in `operation`.
Of course, a case when one needs to use different communication settings on the same bus can be handled with several `spi_config` structures. But this can be an unnecessary waste of memory, and it may be significant in some cases. The whole bus configuration is in `u16_t` field, so it can be easily stored in the driver and compared on the next `transceive` call to see if something actually needs to be reconfigured in the controller. The need of reconfiguration must be checked also in the current version, because each `transceive` call may be done with a different `spi_config` structure than the previous one, and it may (but also may not) contain a completely different configuration.
The similar story is with forcing usage of a separate `spi_config` structure for each device on the bus. When there are many slaves on the bus and all use the same communication settings, one could use one `spi_config` structure and change only the `cs` (or `slave`) field in it accordingly. And I can't see a strong reason why to forbid this.

tbursztyka replied:
Currently configuration change verification is only done on pointer comparison.
I understand your point, but apart from very highly specific use case: nobody is going to change one spi_config operation to access different devices on the bus. 
You are not going to be able to use one spi_config for different drivers, you are pointing out the slave part, but it works the same for master: when you do a driver for X on subsystem K you have your own spi_config. At built time your board enables Y and Z on subsystems M and N respectively, which also access the same spi bus and therefore have a dedicated spi_config. And let's say they have exactly same configurations bits. How are you going to merge these 3 _properly_ so it's transparent for both the device driver dev and the user who will play with Kconfig/dts, without runtime necessary functions etc...? We had this idea in the very beginning of Zephyr and it just ended up impracticable path: too much hassle for such a little tiny gain, if any actually. And I am not even putting the case were spi_cs_control would be different for each devices in the picture.

Plus, I am planning to switch all spi_config to const, everywhere so it'll go on ROM and that's it. And I expect all other APIs to follow the same, where relevant. This will need a change in the device driver model, which I am prototyping, so let's see this in near future. That's out of this PR scope.

anangl replied:
I still consider it (the necessity to provide a separate `spi_config` structure for each device on the bus and for each different configuration of the bus) as an unnecessary limitation. But let's keep it as it is now for this PR, and maybe discuss it later in the context of improvements in the device driver model.

anangl commented on drivers/spi/spi_context.h:
```diff
@@ -19,23 +19,32 @@
+
 struct spi_context {
-	struct spi_config *config;
+	const struct spi_config *config;
```
The standard way of using the context is to copy here the pointer passed to `transceive`. But one may decide to call `transceive` with a pointer to a structure that is an automatic variable, i.e. located on the stack. In case of a synchronous operation it is irrelevant, but for an asynchronous one it can be really destructive. This should be at least documented somewhere, in descriptions of asynchronous parts of SPI API probably.

tbursztyka replied:
Indeed it requires to be documented. spi_config should not be located on the stack.

galak commented:
Would be nice if we had a test that did both master & slave mode on different spi controller instances.

Commit:
drivers/spi: Specify options per-port on DW driver

Generalize clock control. Make interrupt policy per-port.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/clock_control: Enable ARC core support on quark_se driver

Clock control can also be done through the ARC core.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Enable port 3 and 4 on DW driver

This will be useful on Quark_SE ARC core which can access x86 core SPI
controller.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Fix typo on parameters type in DW arc regs definitions

s/u32_s/u32t obviously

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/quark_se_c1000_ss: Switch to SPI DW driver

As QMSI driver does not support new SPI API.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/quark_d2000: Switch to SPI DW driver

As QMSI driver does not support new SPI API.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Removing QMSI driver as it does not support new API

Native DW driver is relevantly used instead.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests/spi: Remove excluded boards

These now support the new API through SPI DW native driver.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

anangl committed:
drivers: spi: Add shim for nrfx SPI driver

This adds a translation layer to make the nrfx driver for the legacy
(i.e. without EasyDMA) nRF SPI peripheral accessible via the updated
Zephyr's API of the SPI driver.
Configuration files are already prepared for adding support for SPIM
(Master with EasyDMA) and SPIS (Slave with EasyDMA) peripherals.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
tests: spi_loopback: Add configurations for a few nRF5 boards

This adds to the spi_loopback test board-specific configurations
for the following boards:
- nrf51_pca10028
- nrf52_pca10040
- nrf52840_pca10056

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

Commit:
drivers/ieee802154: Switch MCR20A driver to new SPI API

And adding support for GPIO CS as well.
It looks like the driver could benefit from centralizing all SPI access
into a unique function, the protocol does not seem too convoluted to do
so, like CC2520 or CC1200.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
@MaureenHelm So with this version of the PR you'll be able to test your driver with MCR20A. If you use:

`CONFIG_IEEE802154_MCR20A_GPIO_SPI_CS=y`
`CONFIG_IEEE802154_MCR20A_GPIO_SPI_CS_DRV_NAME="GPIO_3"`
`CONFIG_IEEE802154_MCR20A_GPIO_SPI_CS_PIN=0`

For instance (make sure boards/arm/frdm_k64f/pinmux.c keeps the gpio function on that PTD0 pin)
It will work. 

There an issue, however, which is not related to SPI, with normal driver logging CCA always time outs for some reason... But with:
`CONFIG_SYS_LOG_IEEE802154_DRIVER_LEVEL=4`
CCA works and thus all works (it can then TX etc...)

 @jfischer-phytec-iot can you check this issue on mcr20a? Thanks.

Commit:
drivers/spi: Remove legacy API support from mcux dspi driver

Now that MCR20A supports the new API, legacy support from mcux dspi
driver can be safely removed.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

erwango commented:
@tbursztyka : Is there something preventing merging this PR?

carlescufi commented:
@erwango the 1.11 release

erwango commented:
@carlescufi : ok thanks!

tbursztyka commented:
@erwango  and the fact it's not fully ready, I'll need to backport @anangl nrfx shim driver and fix the hci sample that still uses legacy api.

Commit:
drivers/spi: Simplify how error is forwarded from ISR handler in DW

No need to test it twice then in completed().

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: spi_context lock makes transceive function reentrant in DW

No need to check on busy bit from hardware.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/ethernet: No need of semaphore for spi in enc28j60

SPI API is reentrant.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/ieee802154: No need of semaphore for spi in mcr20a

SPI API is reentrant.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
samples/bluetooth: Move hci_spi to new SPI API

Removing old API usage.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Remove legacy NRF5 master and slave drivers

As legacy SPI API is being removed.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
spi: Remove SPI legacy API

No drivers nor samples are using it anymore.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

anangl commented on tests/drivers/spi/spi_loopback/src/spi.c:
```diff
@@ -51,56 +51,63 @@ static void to_display_format(const u8_t *src, size_t size, char *dst)
-			return -1;
-		}
+	cs->gpio_dev = device_get_binding(CS_CTRL_GPIO_DRV_NAME);
+	if (!cs->gpio_dev) {
```
This function no longer takes the `cs` pointer as a parameter. Thus, these two lines should be updated to:

```
	spi_cs.gpio_dev = device_get_binding(CS_CTRL_GPIO_DRV_NAME);
	if (!spi_cs.gpio_dev) {
```

Commit:
api/spi: Slave transactions will return received frames on success

Unlike master mode which will always return 0 on success.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>
